packages/cdktf-cli/bin/cmds/deploy.ts

```diff
@@ -67,6 +69,13 @@ class Command extends BaseCommand {
         boolean: true,
         desc: 'Select the "refresh only" planning mode, which checks whether remote objects still match the outcome of the most recent Terraform apply but does not propose any actions to undo any changes made outside of Terraform.',
       })
+      .option("terraform-parallelism", {
```

Comment:
Was there a reason to decide for the CLI-flag approach versus the construct approach? To me, the reasoning of per-stack configuration made sense

Reply:
My opinion on this:

I like the general idea of putting such flags into configuration and we should think about this general idea for other flags as well.
The downside I see to such an approach is that this can't be added to the `cdk.tf.json` that we output and therefore would stop users from using the Terraform CLI in the output directory without knowing that they might need to set such a setting on a certain stack.

For this particular flag, we should confirm its common use-case. My guess is that it is usually used on large stacks when running into limits with some cloud provider APIs – to me personally this would not warrant putting it in the stack (but rather might be something folks would put into their CI).
A different angle: If this is something that could be part of the config, of e.g. a TF provider setting, why is it a flag in Terraform and not a setting in HCL itself? Maybe digging into that direction helps us to uncover more here.

Reply:
My bad, I started with the construct approach, and then switched mid-way. I was planning on explaining it, but lost track. Here's how I was thinking about it:

The reason why the construct approach appealed to me in the start was due to multi-stack deploys. From my limited understanding, parallelism would be used in advanced cases, which leads me to think that it's not something that would be needed across multiple stacks.

However, when I looked at the CLI design, it seemed like we always wanted a certain stack to be specified. Initially, as I was starting out by investigating the CLI, it looked like the deploy command would be deploying a single stack, and nothing else. Then, having a flag would look better, and easier to use.

However, looking at the code, it seems like we would also start planning dependencies of the stack we're deploying. Is that correct? If so, indeed, having a construct is probably better because then we might have a single stack which is a dependency that requires parallelism to be set. At first I thought this would be a blocker for the flag approach, since parallelism would be set for all the dependant stacks, but after some thinking, it seemed like the customer could always deploy the dependant stack with the flag set separately.

In the end, however, the main reason for going back to the CLI flag approach was 'where would the customer try to look for it', and it seems that for people who are accustomed to Terraform, they would look at the CLI.

Sorry for the long-winded comment, but that's my story on how I thought about it and why I ended up with this approach.

Reply:
> However, when I looked at the CLI design, it seemed like we always wanted a certain stack to be specified

Not really, we let users pass a stack selector and use minimatch glob syntax to find all stacks relevant, build a dependency graph and deploy as many in parallel as possible if not restricted. `cdktf deploy '*'` deploys all stacks specified.

> In the end, however, the main reason for going back to the CLI flag approach was 'where would the customer try to look for it', and it seems that for people who are accustomed to Terraform, they would look at the CLI.

I can see that point and it's probably okay for now

> My guess is that it is usually used on large stacks when running into limits with some cloud provider APIs – to me personally this would not warrant putting it in the stack (but rather might be something folks would put into their CI).

From my point of view you contradict yourself here a bit: the flag is used for some cloud providers and large stacks, so if you have a stack with another cloud provider that is not effected that you want to deploy through the same deploy command you now have no way to differentiate and every stacks "suffers" from limited parallelism. 

I'm not saying we have to do this stack-wise, but we discussed this offline to be a construct solution so I was curious on why we drifted away from that approach.

Reply:
I was not part of that discussion, unfortunately 😅 I can see the reasoning with different stacks, though 👍

Reply:
> I was not part of that discussion

Yeah it was just an ad-hoc one while pairing, let's all make sure to document stuff like this in a public channel :) 

> I didn't see a lot of possible flags that we would like to expose in the potential construct. What do you think is another good potential flag?

I would agree, seems to be like the first thing that falls into this category

Reply:
> Personally I'm in favor of a more general solution to arbitrary Terraform cli arguments. See #734

I like the `--` separator idea, but I have two thoughts:

- Having two `parallelism` flags mean different things based on position is not a great experience. The current way seems clearer to me.
- Do you have a proposal for another TF CLI flag that we should try the `--` separator out with?

Reply:
I agree that two arguments of the same name only differing by position isn't a great experience; however, having a special case for an argument isn't great either.
I could see supporting `terraform` prefix generically for any Terraform CLI argument, though just using a separator seems easier to use.
Maybe `parallelism` is more common than I realize, but I've personally never needed to use it so I'm hesitant to give it special consideration. Does the Terraform core team perhaps have data on argument usage?

One argument that I've run into needing is `target`.

Reply:
Yeah, I agree, both aren't great ideas, and I too haven't had the need for using parallelism either. 

`target`, is indeed an interesting flag, but I wonder if we'd want to incorporate it within the cdktf cli as a first-class flag instead of just forwarding it to Terraform instead.